### PR TITLE
[train][multimodal][3/3] Trainer changes to extract multi-modal outputs from GeneratorOutput

### PR DESCRIPTION
**Pull request by nithinvc**
## Summary
Final PR addressing #1493 . This addresses the trainer changes required for compatibility with the new generator. 

- Add VLM (vision-language model) training support by propagating `pixel_values` and `image_grid_thw` through the trainer pipeline
- Extract vision inputs from generator output, wrap them in `TensorList` for variable-length handling, and pass them through to the training batch and forward pass
- Update padding logic to handle `TensorList` types by cloning and concatenating entries instead of using tensor padding

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1498" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: skyrl/train/trainer.py [hunk 1]
```diff
@@ -612,6 +612,15 @@ def convert_to_training_input(self, generator_output: GeneratorOutput, uids: Lis
...
         )
 
+        pixel_values = generator_output.get("pixel_values", None)
+        image_grid_thw = generator_output.get("image_grid_thw", None)
+        if pixel_values is not None:
+            assert len(pixel_values) == len(
+                image_grid_thw
+            ), "Number of pixel values should match number of image grid thw"
+            pixel_values = TensorList(pixel_values)
+            image_grid_thw = TensorList(image_grid_thw)
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/train/trainer.py:625]: fair assert

File: skyrl/train/trainer.py [hunk 2]
```diff
@@ -899,18 +910,25 @@ def pad_batch(self, training_input: TrainingInputBatch) -> TrainingInputBatch:
...
                     # ensures that padding tensors don't count towards the loss
+                    additional_dims = tensor.shape[1:]
                     padding_tensor = torch.zeros(pad_size, *additional_dims, dtype=tensor.dtype, device=tensor.device)
+                    new_tensors[key] = torch.cat([tensor, padding_tensor], dim=0)
                 else:
                     # ensures all padding tensors are in a valid format by cloning `pad_size` from the original input
                     # `pad_size` is guaranteed to be smaller than batch_size
                     padding_tensor = tensor[:pad_size].clone()
-                new_tensors[key] = torch.cat([tensor, padding_tensor], dim=0)
+                    new_tensors[key] = torch.cat([tensor, padding_tensor], dim=0)
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/train/trainer.py:935]: good catch

File: skyrl/train/trainer.py [hunk 3]
```diff
@@ -612,6 +612,18 @@ def convert_to_training_input(self, generator_output: GeneratorOutput, uids: Lis
...
+        image_grid_thw = generator_output.get("image_grid_thw", None)
+        if pixel_values is not None:
+            assert (
+                pixel_values is not None and image_grid_thw is not None
+            ), "Both pixel_values and image_grid_thw must exist for multi-modal inputs"
+            assert len(pixel_values) == len(
+                image_grid_thw
+            ), "Number of pixel values should match number of image grid thw"
+            pixel_values = TensorList(pixel_values)
+            image_grid_thw = TensorList(image_grid_thw)
```
[the remark this answers is not in the document]
Comment (nithinvc) [reply skyrl/train/trainer.py:625]: added the assert, we should always have image_grid_thw with pixel_values

**Review by SumanthRH — APPROVED** [review]
LGTM

**Comment by SumanthRH** [comment]
@nithinvc I will get this PR in first given pending changes in #1486

File: skyrl/train/trainer.py
Comment (CharlieFRuan) [inline skyrl/train/trainer.py:904]: @nithinvc Out of curiosity, did you run into this `pad_batch()` codepath? I have the impression that this is only needed for step-wise training in which the batch_size in `TrainingInputBatch` is not deterministic. Is that also the case for VLM training, or you are mostly supporting step-wise training for VLM training?

Comment (nithinvc) [reply skyrl/train/trainer.py:904]: No not directly - this is for future step-wise support which seemed straightforward to add in this pr. Agree, the non-step-wise training skips this codepath

Comment (CharlieFRuan) [reply skyrl/train/trainer.py:904]: got it, thanks for clarifying!
